src/SSHLibrary/abstractclient.py

```diff
@@ -442,6 +442,7 @@ def _read_until(self, matcher, expected, timeout=None):
             output += self.read_char()
             if matcher(output):
                 return output
+            time.sleep(.00001) # Release GIL so paramiko I/O thread can run
```

Comment:
I think `time.sleep(0)` is the common way to handle this.

Reply:
Changed that.

Reply:
@pekkaklarck Actually doing that reverted the original "faulty" behavior. Any suggestions?

Reply:
I created a paramiko SSH server starting from one of the examples that can be used to reproduce the issue. See: https://github.com/robotframework/SSHLibrary/issues/151#issuecomment-631461271.

When using `time.sleep(0)` on this reproduction scenario, it does succesfully complete, but I've seen it taking between 0m2.608s and 0m9.321s. On the real SSH server I could not let it finish within the 10s timeout.

Whereas with `time.sleep(.00001)`, the highest I got was 0m1.092s.

Reply:
I did more testing:

![table](https://user-images.githubusercontent.com/35524902/82813471-6af55480-9e95-11ea-98ca-f75ee8a8dded.PNG)
![chart](https://user-images.githubusercontent.com/35524902/82813479-6f217200-9e95-11ea-9617-8fb2e96bff95.PNG)

The testing was done against the script I provided for reproduction (not against the real box I have) (for script, see [here](https://github.com/robotframework/SSHLibrary/issues/151#issuecomment-631461271))

For each value I used for testing, I did 10 runs. The values used for the `time.sleep()` are `0`, `.001`, `.0001`, `.00001`, `.000001`, `.000000001`, `.000000000001` and ` ` (i.e. original code; no `time.sleep()`).

The table show the % success rate as well as the average execution time (with 10s timeout).

Reply:
Hi @pekkaklarck, any thoughts on this?

Reply:
Isn't it @mihaiparvu nowadays more on charge of this lib?

Reply:
Hi @mihaiparvu, could you review this?

Reply:
Thanks @freddebacker for your in-depth analysis. I will try this week or next week to reproduce this issue you're having. And see if I find a different solution to this problem other than `sleep()`.

Reply:
Hi @freddebacker, i tried this time.sleep(.00001) in abstractclient.py but doesn't seems to work for me.

Reply:
Hi @mihaiparvu ,

Were you able to reproduce and/or find an alternative solution?
Happy to help if you need more info in reproducing the issue...

Regards,
Fred

Reply:
Hello, I couldn't reproduce the issue. I've used your server script and several Python 3 configurations but the test executed instantly all the time. What paramiko version you are using?

Reply:
@mihaiparvu, if the response output is huge say (5 pages) this is getting hit for me too. probably let me make up a small robot script and attach it here, hope slow response from my side

Reply:
> 
Hi @mihaiparvu ,
> 
> Hello, I couldn't reproduce the issue. I've used your server script and several Python 3 configurations but the test executed instantly all the time. What paramiko version you are using?

Could you confirm that you tested on a multi-core system? The more cores, the worst the results in my observations.

I used these versions:
```
(py3venv) [root@centos7 ~]# pip list
Package                   Version
------------------------- -------
bcrypt                    3.1.7
cffi                      1.14.0
cryptography              2.9.2
future                    0.18.2
paramiko                  2.7.1
pip                       20.1
pycparser                 2.20
PyNaCl                    1.3.0
robotframework            3.1.2
robotframework-sshlibrary 3.4.0
scp                       0.13.2
setuptools                46.3.0
six                       1.14.0
textfsm                   1.1.0
wheel                     0.34.2
```

Reply:
Yes, I've tested with both 2 and 4 cores. 

Unfortunately, I cannot reproduce the problem with these exact same versions. The test passes very fast every time. The only difference I see is that you use centos and I've tested with Debian. I'll setup a VM tomorrow to try it out.

Reply:
Hi @mihaiparvu ,
I was able to reproduce this on a freshly installed debian VM as well. See asciicast here: https://asciinema.org/a/344494

Here is the list of things I think you need to reproduce:
- linux multicore machine
- python3

Let me know if there is anything more I can do to help you reproducing the issue. 
Fred

Reply:
> 
> 
>  See asciicast here: https://asciinema.org/a/344494
> 

Let me record that again, as the robot output is not readable in the cast.

Reply:
The cast is still messing the output in some places, but the important pieces are there.
Here is the new cast: https://asciinema.org/a/344504

Fred

Reply:
Sorry for the very late reply. I've been busy with other projects. 

The good news is that I have finally managed to reproduce the issue. Initially I tried with VM on VirtualBox, but I couldn't reproduce it this way even by configuring the number of CPU cores. I had to install Debian directly on my personal PC and now it's reproduceable.

The bad news is that the problem comes from paramiko, the `recv_ready()` method always returns `False` in a multicore envorinment for this example. There is already an issue open on their issue tracker https://github.com/paramiko/paramiko/issues/1091 but no much progress going on to fix it.

As Pekka implied, adding `sleep` there is more of a hack than a proper fix. But, I want to ask you @freddebacker if this is a blocker issue to you or you have a workaround for it? If it is indeed blocking I could approve the PR and revert in once the paramiko problem is resolved.

Reply:
Hi,

We could work around by setting high timeout values, but this is undesirable because it requires changes to all our existing code and more important we observed some code to run 40 times slower now we migrated to python3. On the other hand, the .00001 sleep as proposed has no significant impact on the runtime, as detailed in my previous python2, python3 comparison with multiple different sleep values.

However I do believe that the way _read_until runs in an endless loop is problematic given the way python run threads sequentially based on the GIL. Paramiko, which put itself in another thread using the threading library, has no chance receiving data if it's thread never gets scheduled by the python interpreter. In such endless loop waiting for another thread to do something, even a .00001 sleep gives more opportunities to python to schedule the another thread to get it's thing done.

Regards, Fred